lib/dry/system/provider_registrar.rb

```diff
@@ -0,0 +1,200 @@
+      #
+      # @api public
+      def [](provider_name)
+        providers[provider_name]
+      end
+      alias_method :provider, :[]
+
+      # @api private
+      def key?(provider_name)
+        providers.key?(provider_name)
+      end
+
+      # Returns a provider if it can be found or loaded, otherwise nil
+      #
+      # @return [Dry::System::Provider, nil]
+      #
+      # @api private
+      def find_and_load_provider(name)
+        name = name.to_sym
+
+        if (provider = providers[name])
+          return provider
+        end
+
+        return if finalized?
+
+        require_provider_file(name)
+
+        providers[name]
+      end
+
+      # @api private
+      def start_provider_dependency(component)
+        if (provider = find_and_load_provider(component.root_key))
+          start(provider.name)
+        end
+      end
+
+      # Returns all provider files within the configured provider_paths.
+      #
+      # Searches for files in the order of the configured provider_paths. In the case of multiple
+      # identically-named boot files within different provider_paths, the file found first will be
+      # returned, and other matching files will be discarded.
+      #
+      # This method is public to allow other tools extending dry-system (like dry-rails)
+      # to access a canonical list of real, in-use provider files.
+      #
+      # @see Container.provider_paths
+      #
+      # @return [Array<Pathname>]
+      # @api public
+      def provider_files
+        @provider_files ||= provider_paths.each_with_object([[], []]) { |path, (provider_files, loaded)| # rubocop:disable Layout/LineLength
+          files = Dir["#{path}/#{RB_GLOB}"].sort
+
+          files.each do |file|
+            basename = File.basename(file)
+
+            unless loaded.include?(basename)
+              provider_files << Pathname(file)
+              loaded << basename
+            end
+          end
+        }.first
+      end
+      deprecate :boot_files, :provider_files
+
+      # @api private
+      def finalize!
+        provider_files.each do |path|
+          load_provider(path)
+        end
+
+        providers.values.each do |provider|
+          start(provider)
+        end
+
+        freeze
+      end
+
+      # @!method finalized?
+      #   Returns true if the booter has been finalized
+      #
+      #   @return [Boolean]
+      #   @api private
+      alias_method :finalized?, :frozen?
+
+      # @api private
+      def shutdown
+        providers.values.each do |provider|
+          stop(provider)
+        end
+      end
+
+      # @api private
+      def prepare(name_or_provider)
+        with_provider(name_or_provider) do |provider|
+          provider.prepare
+          self
+        end
+      end
+
+      # @api private
+      def start(name_or_provider)
+        with_provider(name_or_provider) do |provider|
+          provider.start
+          self
```

Comment:
Shouldn't `self` be below the block?

Reply:
This was how I found it, but I agree it'd be clearer at the very bottom!